drivers/nvdimm/pmem.c

```diff
@@ -45,11 +65,352 @@ static struct nd_region *to_region(struct pmem_device *pmem)
+			goto done;
+		dma_unmap_resource(tr->dma_dev2, tr->raddr, size, dir, 0);
+	}
+
+	/* Unmap BIO data. */
+	dir = op_is_write(bio_op(tr->bio)) ? DMA_TO_DEVICE : DMA_FROM_DEVICE;
+	bio_for_each_bvec(bvec, tr->bio, iter) {
+		if (dma_mapping_error(tr->dma_dev1, tr->addr[i]))
+			break;
+		dma_unmap_page(tr->dma_dev1, tr->addr[i], bvec.bv_len, dir);
+		i++;
+		if (tr->single)
+			continue;
+		if (dma_mapping_error(tr->dma_dev2, tr->addr[i]))
+			break;
+		dma_unmap_page(tr->dma_dev2, tr->addr[i], bvec.bv_len, dir);
+		i++;
+	}
+done:
+	kfree(tr);
+}
+
+static void pmem_dma_callback(void *data, const struct dmaengine_result *result)
+{
+	struct pmem_dma_tr *tr = data;
+	struct bio *bio = tr->bio;
+	struct device *dev = to_dev(tr->pmem);
+
+	if (result->result != DMA_TRANS_NOERROR) {
+		dev_err(dev, "DMA error %x\n", result->result);
+		if (result->result == DMA_TRANS_ABORTED)
+			bio->bi_status = BLK_STS_TRANSPORT;
+		else
+			bio->bi_status = BLK_STS_IOERR;
+	}
+	if (refcount_dec_and_test(&tr->inprog)) {
+		if (tr->do_acct)
+			bio_end_io_acct(bio, tr->start);
+		pmem_dma_unmap(tr);
+		bio_endio(bio);
+	}
+}
+
+static bool pmem_dma_submit_bio(struct pmem_device *pmem, struct bio *bio,
+    bool do_acct, unsigned long start)
+{
+	struct device *dev = to_dev(pmem);
+	struct dma_chan	*dma_chan1, *dma_chan2;
+	struct dma_device *dma_dev1, *dma_dev2;
+	struct dma_async_tx_descriptor *tx;
+	dma_async_tx_callback_result cb;
+	enum dma_data_direction dir;
+	dma_cookie_t cookie, last_cookie1 = 0, last_cookie2 = 0;
+	struct pmem_dma_tr *tr;
+	phys_addr_t laddr, raddr;
+	struct bio_vec bvec;
+	struct bvec_iter iter;
+	unsigned long flags;
+	unsigned int dmas, i, vecs = 0;
+	sector_t sector;
+	static struct dmaengine_result dummy_result = {
+		.result = DMA_TRANS_ABORTED,
+		.residue = 0
+	};
+
+	/* For small I/Os softwate copy is faster. */
+	if (bio->bi_iter.bi_size < min_dma_size)
+		return false;
+
+	/* Choose DMA channels for local and remote PMEM accesses. */
+	laddr = pmem->phys_addr + pmem->data_offset;
+	raddr = op_is_write(bio_op(bio)) ? pmem->rphys_addr : 0;
+	if (raddr)
+		raddr += pmem->data_offset;
+	if (!raddr && pmem->rdma_for_single) {
+		dma_chan1 = pmem->rdma_chan;
+		dma_chan2 = pmem->dma_chan;
+	} else {
+		dma_chan1 = pmem->dma_chan;
+		dma_chan2 = pmem->rdma_chan;
+	}
+	if (dma_chan1 == NULL || dma_chan2 == NULL)
+		return false;
+
+	/* Check the addresses alignment fit DMA device(s) requirements. */
+	dma_dev1 = dma_chan1->device;
+	dma_dev2 = dma_chan2->device;
+	bio_for_each_bvec(bvec, bio, iter) {
+		if (!is_dma_copy_aligned(dma_dev1, bvec.bv_offset,
+		    laddr + iter.bi_sector * 512, bvec.bv_len))
+			return false;
+		if (raddr &&
+		    !is_dma_copy_aligned(dma_dev2, bvec.bv_offset,
+		    raddr + iter.bi_sector * 512, bvec.bv_len))
+			return false;
+		vecs++;
+	}
+
+	/* Collect information needed to complete BIO on DMA completion. */
+	dmas = raddr ? 2 : 1;
+	tr = kmalloc(offsetof(struct pmem_dma_tr, addr[vecs * dmas]),
```

Comment:
Just wondering how much memory it would allocate since we are passing ```offsetof()``` instead of size.

Reply:
Just enough to store vecs * dmas elements of variable-sized addr[] array.  This syntax is actually a GCC'ism, but Clang supports it too.